src/language-advanced.adoc

```diff
@@ -1199,12 +1199,701 @@ TBD
 === Core protocols
 anchor:core-protocols[]
 
+One of the greatest qualities of the core ClojureScript functions is that they are
+implemented around protocols. This makes them open to work on any type that we
+extend with such protocols, be it defined by us or a third party.
+
+
+==== Functions
+
+As we've learned in previous chapters not only functions can be invoked. Vectors are
+functions of their indexes, maps are functions of their keys and sets are functions
+of their values.
+
+We can extend types to be callable as functions implementing the `IFn` protocol. A
+collection that doesn't support calling it as a function is the queue, let's
+implement `IFn` for the `PersistentQueue` type so we're able to call queues as functions
+of their indexes:
+
+[source, clojure]
+----
+(extend-type PersistentQueue
+  IFn
+  (-invoke
+    ([this idx]
+      (nth this idx))))
+
+(def q #queue[0 1 2])
```

Comment:
This example might be clearer if the keys did not equal their corresponding values.